Mathlib/RingTheory/HopkinsLevitzki.lean

```diff
@@ -0,0 +1,110 @@
+/-
+Copyright (c) 2025 Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Junyan Xu
+-/
+import Mathlib.Algebra.Module.Torsion
+import Mathlib.RingTheory.FiniteLength
+import Mathlib.RingTheory.Noetherian.Nilpotent
+import Mathlib.RingTheory.Spectrum.Prime.Noetherian
+
+/-!
+## The Hopkins–Levitzki theorem
+
+## Main results
+
+* `IsSemiprimaryRing.isNoetherian_iff_isArtinian`: the Hopkins–Levitzki theorem, which states
+  that for a module over a semiprimary ring (in particular, an Artinian ring),
+  `IsNoetherian` is equivalent to `IsArtinian` (and therefore also to `IsFiniteLength`).
+
+* In particular, for a module over an Artinian ring, `Module.Finite`, `IsNoetherian`, `IsArtinian`,
+  and `IsFiniteLength` are all equivalent (`IsArtinianRing.tfae`),
+  and a (left) Artinian ring is also (left) Noetherian.
+
+* `isArtinianRing_iff_isNoetherianRing_krullDimLE_zero`: a commutative ring is Artinian iff
+  it is Noetherian with Krull dimension at most 0.
+
+## Reference
+
+* [F. Lorenz, *Algebra: Volume II: Fields with Structure, Algebras and Advanced Topics*][Lorenz2008]
+-/
+
+variable {R M : Type*} [Ring R] [AddCommGroup M] [Module R M]
+
+open Ideal in
+theorem IsSemiprimaryRing.isNoetherian_iff_isArtinian [IsSemiprimaryRing R] :
+    IsNoetherian R M ↔ IsArtinian R M := by
+  have ⟨ss, n, hn⟩ := (isSemiprimaryRing_iff R).mp ‹_›
+  set Jac := Ring.jacobson R
+  replace hn : Jac ^ n ≤ Module.annihilator R M := hn ▸ bot_le
+  induction' n with n ih generalizing M
+  · rw [Submodule.pow_zero, one_eq_top, top_le_iff, Module.annihilator_eq_top_iff] at hn
+    constructor <;> infer_instance
+  obtain _ | n := n
+  · rw [Submodule.pow_one, ← SetLike.coe_subset_coe,
+      ← Module.isTorsionBySet_iff_subset_annihilator] at hn
+    let _ := hn.module
+    have := hn.isSemisimpleModule_iff.mp inferInstance
+    exact IsSemisimpleModule.finite_tfae.out 1 2
+  let N := Jac ^ (n + 1) • (⊤ : Submodule R M)
+  simp_rw [iff_iff_eq] at ih -- otherwise `rw [ih]` below fails!
+  rw [isNoetherian_iff_submodule_quotient N, isArtinian_iff_submodule_quotient N, ih, ih]
+  · rw [← SetLike.coe_subset_coe, ← Module.isTorsionBySet_iff_subset_annihilator,
+      Module.isTorsionBySet_quotient_iff]
+    intro m i hi; exact Submodule.smul_mem_smul hi trivial
+  · rw [← Submodule.annihilator_top, Submodule.le_annihilator_iff, Ideal.IsTwoSided.pow_succ,
+      Submodule.mul_smul, ← Submodule.le_annihilator_iff] at hn
+    exact (Ideal.pow_le_self n.succ_ne_zero).trans hn
+
+variable (R M)
+
+@[stacks 00JB "A ring is Artinian if and only if it has finite length as a module over itself.
+Any such ring is both Artinian and Noetherian."]
```

Comment:
This comment gives an equivalent condition for a ring to be artinian, but the theorem below *assumes* that `R` is artinian...

Reply:
The "if" part holds more generally by [isFiniteLength_iff_isNoetherian_isArtinian](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/FiniteLength.html#isFiniteLength_iff_isNoetherian_isArtinian). How about
```suggestion
@[stacks 00JB "A ring is Artinian if and only if it has finite length as a module over itself.
(For the "if" part, see `isFiniteLength_iff_isNoetherian_isArtinian`.)
Any such ring is both Artinian and Noetherian."]
```

Reply:
My reading of the lean statement is more something like
```suggestion
/-- For a module over an Artinian ring, the following properties are equivalent:
- being finite (as module)
- being Noetherian
- being Artinian
- being of finite length -/
```

Reply:
Sorry, I am completely confused. That stacks tag is proving a lemma about rings, right? Whereas this mathlib theorem is proving a lemma about modules.
Aren't they just talking about different things?

Reply:
I think the only nontrivial part in the quoted statement is that an Artinian ring is finite length as a module over itself, and you can derive this from either the Artinian<->FiniteLength equivalence or the Module.Finite<->FiniteLength equivalence. But I agree the logic is a bit involved. I propose to move this Stacks tag to the next instance that says IsArtinianRing->IsNoetherianRing, and add a comment that reminds the reader that FiniteLength<->Noetherian+Artinian.

Reply:
Sounds good to me

Reply:
In [e2d1b4b](https://github.com/leanprover-community/mathlib4/pull/21451/commits/e2d1b4b0aada1da091779ae34bf32d70286b4a17) I added a lemma that corresponds exactly to the first sentence of 00JB. For the NoetherianRing to ArtinianRing instance, I intend to refer to the second sentence, but the reader won't know what "any such ring" refers to unless the first sentence is also included, so what I do is I include both sentences and make the second boldface. Please let me know if you like the commit.

>  the missing instance from IsArtinianRing to DiscreteTopology (PrimeSpectrum _)

Oh it's already there: [PrimeSpectrum.instDiscreteTopology](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Spectrum/Prime/Noetherian.html#PrimeSpectrum.instDiscreteTopology)

Reply:
Yeah, this looks good to me